### PR TITLE
added more IM model support. Bump to MSBooster v1.2.58

### PR DESCRIPTION
**Pull request by yangkl96**
- find best IM model search implemented with parameter findBestImModel
- propagates delta IM score to pepxml files
- removed FragPipe's check for IM data in favor of MSBooster's built in check. I used an uncalibrated mzml generated from .d and IM scores were turned off due to FragPipe's check
- updated MSBooster

### CODE REVIEW

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java [hunk 1]
```diff
@@ -105,8 +106,8 @@ public boolean configure(Component comp,
       return false;
     }
 
-    if (!predictRT && !predictSpectra) {
-      SwingUtils.showErrorDialog(comp, "At least one of <b>Predict RT</b> or <b>Predict spectra</b> must be enabled if you want to run MSBooster.", NAME + " parameter error");
+    if (!predictRT && !predictSpectra && !predictIm) {
+      SwingUtils.showErrorDialog(comp, "At least one of <b>Predict RT</b> or <b>Predict spectra</b> or <b>Predict IM</b> must be enabled if you want to run MSBooster.", NAME + " parameter error");
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:110]: Is it meaningful only having `predictIm=true`?

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java [hunk 2]
```diff
@@ -143,14 +144,6 @@ public boolean configure(Component comp,
...
     }
 
-    boolean hasTimsTof = false;
-    for (InputLcmsFile t : lcmsToFraggerPepxml.keySet()) {
-      if (t.getPath().getFileName().toString().endsWith(".d")) {
-        hasTimsTof = true;
-        break;
-      }
-    }
-
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:153]: Why removing this check?

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java [hunk 3]
```diff
@@ -162,15 +155,15 @@ public boolean configure(Component comp,
         bufferedWriter.write("renamePin = 1\n");
         bufferedWriter.write("useRT = " + (predictRT ? "true" : "false") + "\n");
         bufferedWriter.write("useSpectra = " + (predictSpectra ? "true" : "false") + "\n");
-        bufferedWriter.write("useIM = " + ((hasTimsTof && predictIm) ? "true" : "false") + "\n");
+        bufferedWriter.write("useIM = " + (predictIm ? "true" : "false") + "\n");
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:158]: There is no need to set `useIM = true` for non-timsTOF data, right?

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java [hunk 4]
```diff
@@ -162,15 +155,15 @@ public boolean configure(Component comp,
...
         bufferedWriter.write("fragger = " + fraggerParams + "\n");
         bufferedWriter.write("deletePreds = false\n"); // FragPipe-PDV need the prediction files.
         bufferedWriter.write("rtModel = " + rtModel + "\n");
         bufferedWriter.write("spectraModel = " + spectraModel + "\n");
         bufferedWriter.write("imModel = " + imModel + "\n");
         bufferedWriter.write("findBestRtModel = " + (findBestRtModel ? "true" : "false") + "\n");
         bufferedWriter.write("findBestSpectraModel = " + (findBestSpectraModel ? "true" : "false") + "\n");
-        bufferedWriter.write("findBestImModel = " + ((hasTimsTof && findBestImModel) ? "true" : "false") + "\n");
+        bufferedWriter.write("findBestImModel = " + (findBestImModel ? "true" : "false") + "\n");
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:166]: There is no need to set useIM = true for non-timsTOF data, right?

File: MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java [hunk 5]
```diff
@@ -173,6 +173,12 @@ private static StringBuilder handle_search_hit(final List<String> searchHit, fin
         if (!Float.isNaN(nttNmc.RTscore)) {
             sb.append(String.format("<search_score name=\"rtscore\" value=\"%f\"/>\n", nttNmc.RTscore));
         }
+        if (!Float.isNaN(nttNmc.ionMobility)) {
+            sb.append(String.format("<search_score name=\"ionmobility\" value=\"%f\"/>\n", nttNmc.ionMobility));
+        }
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java:178]: I think we already have ion mobility in the pepxml file.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java [hunk 6]
```diff
@@ -317,6 +323,8 @@ public static void percolatorToPepXML(final Path pin, final String basename, fin
             final int indexOf_expRT = colnames.indexOf("retentiontime");
             int indexOf_spectralSimilarity = -1;
             int indexOf_predRT = -1;
+            int indexOf_ionMobility = -1;
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java:326]: I think we already have ion mobility in the pepxml file.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java [hunk 7]
```diff
@@ -326,6 +334,12 @@ public static void percolatorToPepXML(final Path pin, final String basename, fin
             if (colnames.contains("delta_RT_loess")) {
                 indexOf_predRT = colnames.indexOf("pred_RT_real_units");
             }
+            if (colnames.contains("ion_mobility")) {
+                indexOf_ionMobility = colnames.indexOf("ion_mobility");
+            }
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java:339]: I think we already have ion mobility in the pepxml file.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java [hunk 8]
```diff
@@ -345,7 +359,16 @@ public static void percolatorToPepXML(final Path pin, final String basename, fin
                     RTscore = Math.abs(Float.parseFloat(split[indexOf_predRT]) -
                             Float.parseFloat(split[indexOf_expRT]));
                 }
-                pinSpectrumRankNttNmc.computeIfAbsent(specId, e -> new NttNmc[max_rank])[rank - 1] = new NttNmc(ntt, nmc, spectralSimilarity, RTscore);
+                float ionMobility = Float.NaN;
+                if (indexOf_ionMobility != -1) {
+                    ionMobility = Float.parseFloat(split[indexOf_ionMobility]);
+                }
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java:365]: I think we already have ion mobility in the pepxml file.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java [hunk 9]
```diff
@@ -1422,7 +1422,8 @@ private static boolean configureTaskGraph(JComponent parent, Path wd, Path jarPa
             msBoosterPanel.findBestImModel(),
             msBoosterPanel.koinaUrl(),
             msBoosterPanel.rtTestModels(),
-            msBoosterPanel.spectraTestModels());
+            msBoosterPanel.spectraTestModels(),
+            msBoosterPanel.imTestModels());
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java:1426]: Is it used anywhere?

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java [hunk 10]
```diff
@@ -73,29 +73,30 @@ public String getCmdName() {
...
+                           String rtModel,
+                           String spectraModel,
+                           String imModel,
+                           boolean findBestRtModel,
+                           boolean findBestSpectraModel,
+                           boolean findBestImModel,
+                           String koinaURL,
+                           String testRtModels,
+                           String testSpectraModels,
+                           String testImModels) {
```
Comment (fcyu) [inline MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:99]: Is it used anywhere?

**Review by fcyu — CHANGES_REQUESTED** [review]

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 1 above)
Comment (fcyu) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:110]: > We can enforce that one of spectra or RT is predicted

That is what the original code is doing.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 2 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:153]: I tested with an uncalibrated mzml produced by msfragger from .d folder and IM was always turned off, even though MSBooster could still extract IM data from the mzml. MSBooster also turns off useIM if it cannot find IM data for a scan

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 3 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:158]: Refer to discussion on uncalibrated mzml

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 4 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:166]: Refer to discussion on uncalibrated mzml

File: MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java (hunk 5 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/tools/percolator/PercolatorOutputToPepXML.java:178]: We do, sending another commit

File: MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java (hunk 9 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java:1426]: Yes, DIA-NN or AlphaPeptDeep can both predict IM/CCS

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 2 above)
Comment (fcyu) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:153]: The uncalibrated mzml is not supposed to be used as the FragPipe input. It is more like an intemediate file.

We also not recommend converting .d to mzML.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 3 above)
Comment (fcyu) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:158]: The uncalibrated mzml is not supposed to be used as the FragPipe input. It is more like an intemediate file.

We also not recommend converting .d to mzML.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 4 above)
Comment (fcyu) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java:166]: The uncalibrated mzml is not supposed to be used as the FragPipe input. It is more like an intemediate file.

We also not recommend converting .d to mzML.

File: MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java (hunk 9 above)
Comment (fcyu) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java:1426]: I am not sure if there is any miscommunication. It seems that MSBooster does not have a `imSearchModelsString` parameter similar to spectral intensity and RT. Your commits also didn't add this option. Thus, the `msBoosterPanel.imTestModels()` is never used.

**Comment by fcyu** [comment]
@yangkl96 why does this pull request have so many unrelated commits and merges?

File: MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java (hunk 9 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java:1426]: I think it is being set properly when I run it

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 1 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java]: added back in latest commit

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 2 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java]: added back in latest commit

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 3 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java]: added back in latest commit

File: MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java (hunk 4 above)
Comment (yangkl96) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/cmd/CmdMSBooster.java]: added back in latest commit

**Comment by yangkl96** [comment]
I forgot to discard the older commits when I synced my fork. They are from previous pull requests that were accepted. I can try opening new pull request if that is cleaner

File: MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java (hunk 9 above)
Comment (fcyu) [reply MSFragger-GUI/src/com/dmtavt/fragpipe/FragpipeRun.java:1426]: OK, you added the option in https://github.com/Nesvilab/FragPipe/pull/1805/commits/bbbbe572d7fafd9c56982dc887fb829dfa2b76bf and https://github.com/Nesvilab/FragPipe/pull/1805/commits/2964fd39e47c317c8a88e7cc08d0d35bab326962. It was not there before and you never told me that you have this option.

**Comment by fcyu** [comment]
Please don't open another new pull request. I will deal with the messy commits of this one.

**Review by fcyu — APPROVED** [review]

**Pull request merged**